### PR TITLE
[CRITICAL] Pipenv-related improvements

### PR DESCRIPTION
**Pull request by frenzymadness**
See the comments and the commit messages for better understanding of what I am trying to achieve.

Part one is for cases when a customer mirror contains also pipenv and has no internet access during a build. This is based on a request from a customer.

Part two is a better and testable solution for pipenv. We basically allow our customers to pin whatever version of pipenv works for them.

### CODE REVIEW

**Comment by frenzymadness** [comment]
It seems that I also need to update our tests because logs are full of errors but the CI is green :disappointed:

**Comment by frenzymadness** [comment]
It seems to be working now. I'm gonna remove the last commit and all tests should pass.

**Comment by frenzymadness** [comment]
Marking this PR as critical because it also fixes our tests.

**Comment by frenzymadness** [comment]
The problem with RHEL 8 is caused by incompatibility between micropipenv and pip 19.2.3. This should be fixed soon by either of these:
* update of RHEL 8 container — we already updated pip in Python 3.8 module to version 19.3.1 which is compatible with micropipenv but it seems to not be available yet
* fix of the incompatibility in micropipenv — https://github.com/thoth-station/micropipenv/pull/91

This PR is ready for a review @torsava

**Comment by frenzymadness** [comment]
A new version of micropipenv is released which should make also rhel8 green. Will see. [test]

File: src/s2i/bin/assemble [hunk 1]
```diff
@@ -36,7 +36,15 @@ function install_tool() {
...
   VENV_DIR=$HOME/.local/venvs/$1
   virtualenv_bin "$VENV_DIR"
-  $VENV_DIR/bin/pip --isolated install -U $1$2  # Combines package name with [extras] if [extras] is defined as $2
+  # First, try to install the tool without --isolated which means that if you
+  # have your own PyPI mirror, it will take it from there. If this try fails, try it
+  # again with --isolated which ignores external pip settings (env vars, config file)
+  # and installs the tool from PyPI (needs internet connetion).
+  $VENV_DIR/bin/pip install -U $1$2  # Combines package name with [extras] if [extras] is defined as $2
+  if [ $? -ne 0 ]; then
+    echo "WARNING: Installation of $1 failed, trying again from official PyPI with pip --isolated"
```
Comment (torsava) [inline src/s2i/bin/assemble:45]:
```suggestion
    echo "WARNING: Installation of $1 failed, trying again from official PyPI with pip --isolated install"
```

File: src/s2i/bin/assemble [hunk 2]
```diff
@@ -36,7 +36,15 @@ function install_tool() {
   {% endif %}
   VENV_DIR=$HOME/.local/venvs/$1
   virtualenv_bin "$VENV_DIR"
-  $VENV_DIR/bin/pip --isolated install -U $1$2  # Combines package name with [extras] if [extras] is defined as $2
+  # First, try to install the tool without --isolated which means that if you
+  # have your own PyPI mirror, it will take it from there. If this try fails, try it
+  # again with --isolated which ignores external pip settings (env vars, config file)
+  # and installs the tool from PyPI (needs internet connetion).
+  $VENV_DIR/bin/pip install -U $1$2  # Combines package name with [extras] if [extras] is defined as $2
```
Comment (torsava) [inline src/s2i/bin/assemble:43]:
```suggestion
  $VENV_DIR/bin/pip install -U $1$2  # Combines package name with [extras] or version specifier if is defined as $2```

File: src/s2i/bin/assemble [hunk 3]
```diff
@@ -36,7 +36,15 @@ function install_tool() {
...
+  if [ $? -ne 0 ]; then
+    echo "WARNING: Installation of $1 failed, trying again from official PyPI with pip --isolated"
+    $VENV_DIR/bin/pip --isolated install -U $1$2  # Combines package name with [extras] if [extras] is defined as $2
```
Comment (torsava) [inline src/s2i/bin/assemble:46]:
```suggestion
  $VENV_DIR/bin/pip install -U $1$2  # Combines package name with [extras] or version specifier if is defined as $2```

File: src/README.md [hunk 4]
```diff
@@ -123,12 +123,17 @@ file inside your source code repository.
...
-    Set this variable to use [Pipenv](https://github.com/kennethreitz/pipenv),
+    Set this variable to use [Pipenv](https://github.com/pypa/pipenv),
     the higher-level Python packaging tool, to manage dependencies of the application.
     This should be used only if your project contains properly formated Pipfile
     and Pipfile.lock.{% if spec.version in ["2.7"] %} (Implies `UPGRADE_PIP_TO_LATEST` to satisfy dependencies of
     Pipenv.){% endif %}
 
+* **PIN_PIPENV_VERSION**
+
+    Set this variable together with `ENABLE_PIPENV` to use a specific version of Pipenv.
```
Comment (torsava) [inline src/README.md:134]:
```suggestion
    Set this variable together with `ENABLE_PIPENV` to use a specific version of Pipenv.
    If not set, the latest stable version from PyPI is installed.
```

File: examples/pin-pipenv-version-test-app/app.sh [hunk 5]
```diff
@@ -0,0 +1,20 @@
+#!/bin/bash
+
+echo "Testing PIN_PIPENV_VERSION (set in .s2i/environment) ..."
+pipenv_version=`pipenv --version`
+expected="pipenv, version 2018.11.26"
+if [ "$pipenv_version" != "$expected" ]; then
+  echo "ERROR: pipenv version is different than expected."
```
Comment (torsava) [inline examples/pin-pipenv-version-test-app/app.sh:7]:
```suggestion
  echo "ERROR: pipenv version is different than expected."
  echo "Expected: ${expected}"
  echo "Actual: ${pipenv_version}"
```

**Comment by frenzymadness** [comment]
Thanks @torsava, I'm gonna squash your suggested changes and if the CI will be ok, I'll merge it.

File: src/s2i/bin/assemble (hunk 3 above)
Comment (frenzymadness) [reply src/s2i/bin/assemble]: BTW, I'll adjust this one because the indentation should remain there.

**Comment by torsava** [comment]
Tests are failing on centos7 and fedora for some reason.

File: src/s2i/bin/assemble (hunk 3 above)
Comment (torsava) [reply src/s2i/bin/assemble]: Oh, of course, my mistake.

**Comment by frenzymadness** [comment]
[test]

**Comment by frenzymadness** [comment]
All green :tada:

**Comment by torsava** [comment]
> All green

I see 2 failures again...

**Comment by torsava** [comment]
Hm, I had to refresh 2 times before the new test results loaded.

Anyway, feel free to merge! :)

**Pull request merged**